src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,274 @@
+        {
+            get { return _responseStream; }
+        }
+
+        #endregion
+
+        internal CurlResponseMessage(HttpRequestMessage request)
+        {
+            RequestMessage = request;
+            Content = new StreamContent(_responseStream);
+        }
+    }
+
+        internal sealed class CurlResponseStream : Stream
+        {
+            private volatile bool _disposed = false;
+            private Stream _innerStream;
+            private readonly object _lockObject = new object();
+            private readonly ManualResetEvent _dataRequestedEvent = new ManualResetEvent(false);
+            private readonly ManualResetEvent _dataReadyEvent = new ManualResetEvent(false);
+            private readonly ManualResetEvent _noMoreDataEvent = new ManualResetEvent(false);
+
+            private class UnmanagedMemoryStream : MemoryStream
+            {
+                // TODO: Is UnmanagedMemoryStream available in System.IO 4.0.10?
+                internal unsafe UnmanagedMemoryStream(byte* pointer, long length) : base((int)length)
+                {
+                    for (int i = 0; i < (int)length; i++)
+                    {
+                        base.WriteByte(*(pointer + i));
+                    }
+                    base.Seek(0, SeekOrigin.Begin);
+                }
+            }
+
+            internal CurlResponseStream() : base()
+            {
+            }
+
+            internal void SignalComplete(bool isCancellation)
+            {
+                if (!_disposed)
+                {
+                    _noMoreDataEvent.Set();
+#if DEBUG
+                    if (Logging.On)
+                    {
+                        Logging.PrintInfo(Logging.Http, this,
+                            "SignalComplete: " + (isCancellation ? "Cancellation" : "No more data") + " for " +
+                            Logging.GetObjectLogHash(this));
+                    }
+#endif
+                }
+            }
+
+            internal unsafe void WaitAndSignalReaders(byte* pointer, long length)
+            {
+                CheckDisposed();
+                var events = new ManualResetEvent[] { _dataRequestedEvent, _noMoreDataEvent };
+                var eventIndex = WaitHandle.WaitAny(events, Timeout.InfiniteTimeSpan);
+                // Check for an "out-of-band" cancellation
+                if ((1 == eventIndex) || _noMoreDataEvent.WaitOne(TimeSpan.Zero))
+                {
+                    throw new InvalidOperationException("Cancelled due to no readers");
+                }
+                lock (_lockObject)
+                {
+                    _innerStream = new UnmanagedMemoryStream(pointer, length);
+#if DEBUG
+                    if (Logging.On) Logging.PrintInfo(Logging.Http, this, "ReaderWriter wrote " + Logging.GetObjectLogHash(this) + " count=" + length);
+#endif
+                    _dataRequestedEvent.Reset();
+                    _dataReadyEvent.Set();
+                }
+            }
```

Comment:
Need a wait here to ensure that all data has been read before giving up the pointer